packages/vite/src/node/config.ts

```diff
@@ -829,13 +829,13 @@ export async function loadConfigFromFile(
 
     if (isMjs) {
       const fileUrl = require('url').pathToFileURL(resolvedPath)
+      const bundled = await bundleConfigFile(resolvedPath, true)
+      dependencies = bundled.dependencies
```

Comment:
Why was this code moved up? Maybe it is an artifact from trying to fix the issue?

Reply:
Because the JS configuration file also needs a bundle, and then collects the dependencies, which is consistent with the processing of the TS file. Otherwise, the JS configuration changes will still not be noticed by the server. Is there a better way?

Reply:
The current config loading code tries to use node's default loading mechanism as much as possible, and falls back to bundling only when needed. I don't think it's good to always bundle in order to get the list of dependencies. IMO I'd leave the code as is and have bundling get special features. Related PR: #3031

Reply:
I think it is necessary to ensure the HMR consistency of TS and JS configuration files, and the cost of Esbuild's bundle analysis is negligible, just like doing a dependency analysis before pre-bundle.

Reply:
My question here was about moving `bundleConfigFile` out of the first condition. I don't see how it is used in the else branch.
I'll let @bluwy discuss with you here, as he has more knowledge of how Vite deals with config files. 

@sanyuan0704 the PRs that you are sending lately are great. Thanks a lot for digging into these issues. Just a note, please avoid sending PTAL messages tagging others to bump them, maintainers, and collaborators are going to get to them when they have time. A PTAL message doesn't make a difference and generates notifications for everyone subscribed. Thanks again!

Reply:
There are two options here:

1. Always bundle config files with esbuild regardless if they can already be imported by node directly.
2. Keep the code as-is and use node import whenever possible, falling back to esbuild otherwise.

Currently we're doing no2 at the expense of the #5780 bug. But a caveat if we switch to no1 is that the workaround for #5694 (`.default` for some vite plugins) wouldn't work anymore, and it's a breaking change.

That was the reason I prefer we keep things as is, unless we're comfortable with the breaking change and always running esbuild.

Reply:
@bluwy Thank you for reminding that there is still a workaround problem. I read the issue and tested the vite plugin in the issue on my pr branch. The user's original workaround is still valid. Of course, if I change it according to my current code to take effect.

Reply:
Interesting. If the `.default` workaround still works when bundled by esbuild, then it may be safe then. I guess it now comes down to whether the Vite team is fine with always running esbuild for the config.

Reply:
I don't know if Vite team can accept this treatment. While ensuring safety, it also ensures the consistency of ts and js configuration files in HMR. Anyway, I think it is a good thing for Vite itself. Thanks for the reply again.

Reply:
Still, these first two lines dont need to change, no? I dont see a side effect when bundling, and deps are not used in the else branch. Am I missing something?

Reply:
Now it is necessary to execute the build of esbuild for the else case, because Vite needs to collect the configuration file dependencies.If do not move up, these dependencies will not be collected in the js configuration file.

Reply:
Therefore, the focus is on the collection of dependencies, not whether it is used by else.

Reply:
Uh, `dependencies` isn't a local variable to that scope. Sorry for the noise 🤦🏼‍♂️